src/storage/entry.rs

```diff
@@ -0,0 +1,201 @@
+        if self.entities.is_alive(e) {
+            unsafe {
+                let entries = self.entries();
+                let (_, mut value): (BitSetAll, _) = entries.open();
+                Ok(Entries::get(&mut value, e.id()))
+            }
+        } else {
+            let gen = self
+                .entities
+                .alloc
+                .generation(e.id())
+                .unwrap_or(Generation::one());
+            Err(WrongGeneration {
+                action: "attempting to get an entry to a storage",
+                actual_gen: gen,
+                entity: e,
+            })
+        }
+    }
+
+    /// Returns a `Join`-able structure that yields all indices, returning
+    /// `Entry` for all elements
+    pub fn entries<'a>(&'a mut self) -> Entries<'a, 'e, T, D> {
+        Entries(self)
+    }
+}
+
+/// `Join`-able structure that yields all indices, returning `Entry` for all elements
+pub struct Entries<'a, 'b: 'a, T: 'a, D: 'a>(&'a mut Storage<'b, T, D>);
+
+impl<'a, 'b: 'a, T: 'a, D: 'a> Join for Entries<'a, 'b, T, D>
+where
+    T: Component,
+    D: Deref<Target = MaskedStorage<T>>,
+{
+    type Type = StorageEntry<'a, 'b, T, D>;
+    type Value = &'a mut Storage<'b, T, D>;
+    type Mask = BitSetAll;
+
+    unsafe fn open(self) -> (Self::Mask, Self::Value) {
+        (BitSetAll, self.0)
+    }
+
+    unsafe fn get(value: &mut Self::Value, id: Index) -> Self::Type {
+        // This is HACK. See implementation of Join for &'a mut Storage<'e, T, D> for
+        // details why it is necessary.
+        let storage: *mut Storage<'b, T, D> = *value as *mut Storage<'b, T, D>;
+        if (*storage).data.mask.contains(id) {
+            StorageEntry::Occupied(OccupiedEntry {
+                id,
+                storage: &mut *storage,
+            })
+        } else {
+            StorageEntry::Vacant(VacantEntry {
+                id,
+                storage: &mut *storage,
+            })
+        }
+    }
+
+    #[inline]
+    fn is_unconstrained() -> bool {
+        true
+    }
+}
+
+/// An entry to a storage which has a component associated to the entity.
+pub struct OccupiedEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    id: Index,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+where
+    T: Component,
+    D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Get a reference to the component associated with the entity.
+    pub fn get(&self) -> &T {
+        unsafe { self.storage.data.inner.get(self.id) }
+    }
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+where
+    T: Component,
+    D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Get a mutable reference to the component associated with the entity.
+    pub fn get_mut(&mut self) -> &mut T {
+        unsafe { self.storage.data.inner.get_mut(self.id) }
+    }
+
+    /// Converts the `OccupiedEntry` into a mutable reference bounded by
+    /// the storage's lifetime.
+    pub fn into_mut(self) -> &'a mut T {
```

Comment:
This looks suspicious. Wouldn't this let you peddle the reference to a variable which can then be accessed in a join? Or is there something I am missing here.

Reply:
Ah it was there before. It would be good to have some explanation why it's safe.

Reply:
Full signature for this function is `pub fn into_mut(self: OccupiedEntry<'a, 'b, T, D>) -> &'a mut T`,
so the returned reference can not outlive the entry which can not outlive Entries which can not outlive Storage.
References are mutable, so rust won't allow you to create 2 Entries to the same storage.
However you could misuse __unsafe__ Join::get to create `StorageEntry` more than once, but:
- it is possible for any Join over mutable data
- and Join::get is __unsafe__